src/core/src/op/mod.cpp

```diff
@@ -31,6 +48,182 @@ struct Evaluate : ov::element::NoAction<bool> {
+
+    if (!eq_op.evaluate(output_mask_vector, {m_lb, zero_t})) {
+        return false;
+    }
+
+    auto tmp_mask = TensorVector{{element::boolean, output_mask.get_shape()}};
+    if (!eq_op.evaluate(tmp_mask, {m_ub, zero_t})) {
+        return false;
+    }
+    or_op.evaluate(output_mask_vector, {output_mask_vector.front(), tmp_mask.front()});
+
+    if (!eq_op.evaluate(tmp_mask, {m_lb, max_t})) {
+        return false;
+    }
+    or_op.evaluate(output_mask_vector, {output_mask_vector.front(), tmp_mask.front()});
+
+    if (!eq_op.evaluate(tmp_mask, {v_ub, max_t})) {
+        return false;
+    }
+    return or_op.evaluate(output_mask_vector, {output_mask_vector.front(), tmp_mask.front()});
+}
+
+/**
+ * @brief Get the m (modulo) inputs bound with set to one values which can cause undefined result.
+ *
+ * @param bounds  Modulo operator inputs bounds.
+ * @param mask    Mask with undefined result values.
+ * @return Vector of bounds tensors.
+ */
+TensorVector get_m_bounds(const TensorVector& bounds, const Tensor& mask) {
+    const auto select_op = v1::Select();
+    const auto one_t = ov::util::make_tensor_of_value(bounds.front().get_element_type(), 1);
+
+    auto m_bounds = TensorVector();
+    std::transform(bounds.begin() + 2, bounds.end(), std::back_inserter(m_bounds), [&](const Tensor& b) {
+        auto tmp = TensorVector{{b.get_element_type(), b.get_shape()}};
+        return select_op.evaluate(tmp, {mask, one_t, b}) ? tmp.front() : Tensor{};
+    });
+    return m_bounds;
+}
+
+/**
+ * @brief Evaluate modulo upper or lower bound.
+ *
+ * @param op        Pointer to modulo node.
+ * @param outputs   Tensor vector with one tensor to store bounds result.
+ * @param is_lower  True to evaluate lower otherwise evaluate upper.
+ * @return True if outputs has valid data otherwise false.
+ */
+bool evaluate_bound(const Node* const op, TensorVector& outputs, bool is_lower) {
+    const auto bounds = mod::get_bounds(op);
+
+    if (mod::are_bounds_valid(bounds)) {
+        const auto& v_lb = bounds[0];
+        const auto& v_ub = bounds[1];
+        const auto& in_et = v_lb.get_element_type();
+        const auto in_shape = ov::op::infer_broadcast_shape(op, v_lb.get_shape(), bounds[2].get_shape());
+
+        // Get undefined results mask for modulo.
+        auto undefined_result_mask = Tensor{element::boolean, in_shape};
+        if (!mod::evaluate_undefined_result_mask(undefined_result_mask, bounds)) {
+            return false;
+        }
+
+        // Set 2nd input values to 1 for undefined results mask (0, inf, etc.)
+        const auto m_bounds = mod::get_m_bounds(bounds, undefined_result_mask);
+        if (!mod::are_bounds_valid(m_bounds)) {
+            return false;
+        }
+        // Evaluate bound.
+        outputs[0].set_shape(in_shape);
+        using namespace ov::element;
+        if (!IfTypeOf<i8, i16, i32, i64, u8, u16, u32, u64>::apply<mod::EvaluateBound>(in_et,
```

Comment:
Assuming that the inputs for `evaluate_bounds`, are usually outputs (at some point) of the `ShapeOf`, (i32 and i64) could be sufficient. Also the "smaller" precisions could be converted to the "bigger" (just for the calculations) to avoid template specialization for all of the types.

Reply:
I'm not 100% that precision can be removed. 
Inserting Convert precision will required design graph or make some transformation which insert this convert and it has to be converted twice (one from other to e.g. i64 then back from i64. to source precision). Add it inside evaluate will add additional data copy. 

`ShapeOf` also supports i32, i64, u32, u64 in evaluate but for op creation only i32, i64 are valid.

But if the list can be reduce then can change it.
@jane-intel Do you know if  precisions list can be reduced?